### PR TITLE
zip: fix build with GCC 14

### PR DESCRIPTION
**Pull request by FliegendeWurst**
GCC 14 breaks the package

ref. #356812

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilazy** [comment]
I’d actually like to merge this pretty promptly as I’ve had a few things blocked on OpenJDK builds. But it’s kind of awful that we’re just downgrading the compilation to sweep some UB further under the rug – can you add a comment that `-O3` exposes undefined behaviour with GCC 14, that it manifested in a broken `openjdk8` build, and a TODO to at least pretend that someone will investigate this properly one day? Because it’s always possible that GCC 15 starts exposing it at `-O2`…

**Comment by FliegendeWurst** [comment]
So... I was apparently foolish enough to assume that "no more of the new warnings" = "fixed zip package".
This fix is still not enough. I will investigate further.

There is no maintained fork of zip, right?

**Comment by RossComputerGuy** [comment]
> There is no maintained fork of zip, right?

Afaik, no. Same goes for unzip.

**Comment by emilazy** [comment]
Gentoo has More Patches that might be worth trying: https://gitweb.gentoo.org/repo/gentoo.git/tree/app-arch/zip/files

**Comment by FliegendeWurst** [comment]
The same error occurs at `-O0`, and the Gentoo patches do not help.

**Comment by RossComputerGuy** [comment]
Wouldn't just dropping the optimization flag work?

**Comment by FliegendeWurst** [comment]
> Wouldn't just dropping the optimization flag work?

That would be equivalent to `-O0`, right? Though it looks a bit cleaner.

**Comment by FliegendeWurst** [comment]
I've pushed an alternative approach where we use GCC 13 instead. Not very nice but it works

**Comment by emilazy** [comment]
Not sure I feel good about this – if it’s GCC 14 miscompiling the code then we’re in trouble and can expect other issues throughout the tree. OTOH if it’s UB in `zip` then pinning an old compiler doesn’t seem very sustainable. But maybe it’s the best solution for now…

**Comment by FliegendeWurst** [comment]
I just found https://bugzilla.redhat.com/show_bug.cgi?id=2165653 and will try their patch later. But it doesn't really match our problem.

**Comment by FliegendeWurst** [comment]
Nevermind, we already include that patch.

**Review by emilazy — CHANGES_REQUESTED** [review]
This worries me, but I’m okay proceeding with it for now if you split the patches into another commit and adjust the comment.

**Comment by FliegendeWurst** [comment]
So I just found another patch: https://sources.debian.org/data/main/z/zip/3.0-14/debian/patches/12-fix-build-with-gcc-14.patch
(please advise where I can get that reproducibly, I can't seem to find it in the Debian Salsa)

Which I want to try out. I was comparing the build logs, and this patch removes some differences.

File: pkgs/by-name/zi/zip/package.nix [hunk 1]
```diff
@@ -27,17 +27,45 @@ stdenv.mkDerivation rec {
   ];
 
   patches = [
-    # Trying to use `memset` without declaring it is flagged as an error with clang 16, causing
-    # the `configure` script to incorrectly define `ZMEM`. That causes the build to fail due to
-    # incompatible redeclarations of `memset`, `memcpy`, and `memcmp` in `zip.h`.
-    ./fix-memset-detection.patch
     # Implicit declaration of `closedir` and `opendir` cause dirent detection to fail with clang 16.
     ./fix-implicit-declarations.patch
-    # Buffer overflow on Unicode characters in path names
```
Comment (FliegendeWurst) [inline pkgs/by-name/zi/zip/package.nix:36]: This patch was moved below as part of the Gentoo patches.

**Review by emilazy — APPROVED** [review]
Builds on `aarch64-linux` and `aarch64-darwin`. Thanks for getting to the root of this! Much happier with this than pinning the compiler.

**Pull request merged**